src/instructlab/model/full_train.py

```diff
@@ -0,0 +1,319 @@
+        aggregated_values = torch.zeros(3, dtype=torch.float16).to(dev)
+
+        for step, batch in enumerate(dataloader):
+            aggregated_values[0] = batch.pop("num_loss_counted_tokens")
+            aggregated_values[1] = len(batch["input_ids"])
+
+            # Move and cast batch data to device
+            for k in batch:
+                if total_ram < 16:
+                    if k in ["input_ids", "attention_mask"]:
+                        # these two need to be of type long if using .half()
+                        batch[k] = batch[k].to(device=dev, dtype=torch.long)
+                    else:
+                        batch[k] = batch[k].to(device=dev, dtype=torch.float16)
+                else:
+                    batch[k] = batch[k].to(device=dev)
+
+            output = model(**batch, use_cache=False)
+            loss = output.loss
+            aggregated_values[2] = loss.item()
+
+            num_loss_counted_tokens = aggregated_values[0]
+            loss = loss / num_loss_counted_tokens
+
+            loss = loss / accum  # Scale the loss for accumulation steps
+
+            logger.info(
+                f"\nEpoch: {epoch}, Step: {global_step}, Rank: 0, loss = {loss}"
+            )
+
+            # Gradient accumulation
+            loss.backward()  # Backward pass
+
+            # Clear cache before optimizer step
+            if dev.type == "mps":
+                torch.mps.empty_cache()
+
+            # if we are on a step which is divisible by 4, step and zero gradients
+            if (step + 1) % accum == 0:
+                optimizer.step()  # Optimizer step
+                optimizer.zero_grad()  # Zero gradients
+
+                # Clear cache after optimizer step
+                if dev.type == "mps":
+                    torch.mps.empty_cache()
+
+            global_step += 1
+            inner_pb.update(1)
+
+        # Clear cache at the end of the epoch if needed
+        if dev.type == "mps":
+            torch.mps.empty_cache()
+
+        output_dir = (
+            Path(train_args.data_output_dir)
+            / "hf_format"
+            / f"samples_{(global_step*8)}"
+        )
+
+        logger.info(f"Saving Model to: {output_dir}")
+        model_state = model.state_dict()
+        output_dir.mkdir(parents=True, exist_ok=True)
+        output_model_file = output_dir / "pytorch_model.bin"
+        output_config_file = output_dir / "config.json"
+
+        tmp_conf = copy(model.config)
+
+        torch.save(model_state, str(output_model_file))
+        tmp_conf.to_json_file(str(output_config_file))
+        tokenizer.save_pretrained(str(output_dir))
+
+        # convert to GGUF at the end so llamacpp can serve the model originally in .bin
+        llamacpp_convert_to_gguf.convert_llama_to_gguf(
+            model=output_dir,
+            pad_vocab=True,
+            skip_unknown=True,
+            outfile=os.path.join(output_dir, "pytorch_model.gguf"),
```

Comment:
might be better for us to pick one agreed upon file name for the final trained model and use that across all training loops (simple/mlx/this one) so that users are not confused about what file to expect and whether or not this whole training worked for them

Reply:
so, this gguf is going inside of the `samples_...` folder, I don't think ppl will be confused as to which is which. the legacy train uses `ggml...` I think and library uses the `samples_...` format too with just a .bin inside

If anything, when `linux_train.py` gets re-written I feel like maybe the change should happen there not here?